src/KrakenBattle.cpp

```diff
@@ -0,0 +1,158 @@
+   const std::string emptyUUID = {"00000000-0000-0000-0000-000000000000"};
+   const std::vector<uint8_t> emptyData = {};
+}
+
+
+
+namespace KrakenBattle {
+
+   /**
+   * formats the data to send to the Harpoon according to the specification for
+   *  @ref file and pcap sending
+   * @param uuid
+   * @param type one of Data, Done, Error, End
+   * @param data to send (optional, used for SendType::Data)
+   * @param error message (optional, used for SendType::Error)
+   * @return merged data uuid<SendType>optional:data/optional:error_msg
+   */
+   Kraken::Chunks  MergeData(const std::string& uuid, const KrakenBattle::SendType& type, const Kraken::Chunks& data, const std::string& error_msg) {
+      const std::string sendType = EnumToString(type);
+      const std::string& uuidToSend = (SendType::End == type ? emptyUUID : uuid);
+      std::vector<uint8_t> merged;
+      merged.reserve(uuidToSend.size() + sendType.size() + data.size() + error_msg.size());
+      // uuid, type, possible:data, possible:error
+      std::copy(uuidToSend.begin(), uuidToSend.end(), std::back_inserter(merged));
+      std::copy(sendType.begin(), sendType.end(), std::back_inserter(merged));
+
+      if (SendType::Data == type) {
+         std::copy(data.begin(), data.end(), std::back_inserter(merged));
+      } else if (SendType::Error == type) {
+         std::copy(error_msg.begin(), error_msg.end(), std::back_inserter(merged));
+      }
+      return merged;
+   }
+
+
+   /**
+   *  Send  Chunks over Kraken to a Harpoon
+   *  If chunks to send is more than the @ref Kraken::MaxChunkSizeInBytes() then
+   *  it will split up the sending into several separate sends (@ref Kraken::SendTidalWave)
+   * @param kraken to send over
+   * @param uuid
+   * @param sendState
+   * @param chunk to send (optional content, for SendType::Data)
+   * @param error to send (optional content, for SendType::Error)
+   */
+   Kraken::Battling SendChunks(Kraken* kraken, const std::string& uuid, const KrakenBattle::SendType& sendState, const Kraken::Chunks& chunk, const std::string& error) {
+      const size_t kSplitSize = kraken->MaxChunkSizeInBytes();
+      const auto kTotalToSend = MergeData(uuid, sendState, chunk, error);
+      const size_t kTotalSize = kTotalToSend.size();
+      const Kraken::Chunks kHeader = MergeData(uuid, sendState, {}, {}); // {}: ignored
+      const size_t kSplitSizeAdjusted = kSplitSize - kHeader.size();
+      CHECK(kSplitSize > kHeader.size());
+
+      auto result = Kraken::Battling::CONTINUE;
+
+      if (kTotalSize <= kSplitSize) {
+         result = kraken->SendTidalWave(kTotalToSend);
+      } else {
+
+         std::vector<uint8_t> toSend;
+         toSend.reserve(kSplitSize);
+         toSend.assign(kTotalToSend.begin(), kTotalToSend.begin() + kSplitSize);
+
+         CHECK(toSend.size() == kSplitSize);
+         result = kraken->SendTidalWave(toSend);
+
+         for (size_t i = kSplitSize; i < kTotalSize; i += kSplitSizeAdjusted) {
```

Comment:
Why do we send the first TidalWave, and then send the rest? Can you calculate the number of TidalWaves you will have to send beforehand and then send them all the same way?